FEATURE: [2.2] [AdminBundle] make search menu in admin sidebar component hookable

crydotsnake opened the pull request:
| Q               | A
|-----------------|-----
| Branch?         | 2.2
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #18494,
| License         | MIT

This pull request enhances the extendability of the sidebar component in the Sylius admin UI by allowing the search menu to be disabled via the twig hooks functionality.

This can be archieved like this for example:

```yaml
# config/packages/sylius_admin_twig_hooks.yaml
sylius_twig_hooks:
    enable_autoprefixing: true
    hooks:
      'sylius_admin.common.component.sidebar.menu':
        search:
          enabled: false
```

crydotsnake committed:
FEATURE: [2.2] [AdminBundle] make search menu in admin sidebar component hookable

TASK: apply review feedback

TASK: revert change and switch back too initial way of hook the search

TASK: apply review feedback

TASK: refactor namings

TASK: add priorities too twig hooks

TASK: adjust priorities

Rafikooo commented on src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/common/component/sidebar.yaml:
```diff
@@ -15,3 +15,8 @@ sylius_twig_hooks:
+        'sylius_admin.common.component.sidebar.search_menu':
+            search_menu:
+                template: '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig'
+                priority: 0
```
Putting this above between `toggle_button` and `menu` makes it more concise imo:
```yaml
            toggle_button:
                template: '@SyliusAdmin/shared/crud/common/sidebar/toggle_button.html.twig'
                priority: 100
            search:
                template: '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig'
                priority: 50
            menu:
                template: '@SyliusAdmin/shared/crud/common/sidebar/menu.html.twig'
                priority: 0
```

Rafikooo commented on src/Sylius/Bundle/AdminBundle/templates/shared/crud/common/sidebar/menu.html.twig:
```diff
@@ -1,4 +1,4 @@
 <div class="collapse navbar-collapse" id="sidebar-menu">
-    {% include '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig' %}
+    {% hook 'search_menu' %}
```
```suggestion
```
Then we don’t need to fire the hook here anymore. Moreover, it’s not coupled with an order anymore, so if someone wants to change it (kinda weird to have a search bar below the menu), it works just by adjusting the YAML config :)

crydotsnake commented:
@Rafikooo Thanks for your feedback - i totally agree with you :) Tested it again and it works as expected.

commits are also squashed 🙂

crydotsnake commented:
> Last request, @crydotsnake aka Simon. After the refactor, the search element is no longer wrapped with `<div class="collapse navbar-collapse" id="sidebar-menu">`. Could you verify the purpose of these classes and check if they affect the search input somehow (especially with different screen sizes, etc.)? If so, I think we’re supposed to wrap this menu element as well or think a bit more about the solution.

Good point.

First thing i see is, that the search menu is now always visible in mobile. But before the menu was only visible when open the sidebar.

## locally

<img width="447" height="948" alt="image" src="https://github.com/user-attachments/assets/abadfe58-5848-4cb1-9b66-60ed96c63b91" />

<img width="474" height="971" alt="image" src="https://github.com/user-attachments/assets/4b59de21-b7bc-4f4a-bb74-8aa191ae1be2" />


## Sylius demo (v2.demo.sylius.com)

<img width="441" height="940" alt="image" src="https://github.com/user-attachments/assets/ddc7a569-65dc-40a5-8ef8-1009537b5f43" />

<img width="468" height="977" alt="image" src="https://github.com/user-attachments/assets/3f1e507f-f5a3-428b-82cd-ae20833fe8e7" />


Then i would suggest we wrap it inside the `<div class="collapse navbar-collapse" id="sidebar-menu">` div and hook it like i had it before the refcator change until we have a better solution.

Wdyt?

Rafikooo commented:
Yep, that’s why I reverted my message. I’m afraid we have to drop my proposal and go back to your solution where the search element is inside `<div class="collapse navbar-collapse" id="sidebar-menu">`, so the mobile view still works correctly. From what I’ve found, we can’t have two Bootstrap navbars - it would break the responsive design.

crydotsnake commented:
> Yep, that’s why I reverted my message. I’m afraid we have to drop my proposal and go back to your solution where the search element is inside `<div class="collapse navbar-collapse" id="sidebar-menu">`, so the mobile view still works correctly. From what I’ve found, we can’t have two Bootstrap navbars - it would break the responsive design.

I should have checked that more carefully...

I switched back too my first solution and squashed the commits..

Rafikooo commented on src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/common/component/sidebar.yaml:
```diff
@@ -15,3 +15,8 @@ sylius_twig_hooks:
                 template: '@SyliusAdmin/shared/crud/common/sidebar/logo/image.html.twig'
                 priority: 0
+
+        'sylius_admin.common.component.sidebar.search':
```
```suggestion
        'sylius_admin.common.component.sidebar.menu':
```

Rafikooo commented on src/Sylius/Bundle/AdminBundle/templates/shared/crud/common/sidebar/menu.html.twig:
```diff
@@ -1,4 +1,4 @@
 <div class="collapse navbar-collapse" id="sidebar-menu">
-    {% include '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig' %}
+    {% hook 'search' %}
```
```suggestion
    {% hook 'menu' %}
```

Rafikooo commented on src/Sylius/Bundle/AdminBundle/templates/shared/crud/common/sidebar/menu.html.twig:
```diff
@@ -1,4 +1,4 @@
 <div class="collapse navbar-collapse" id="sidebar-menu">
-    {% include '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig' %}
+    {% hook 'search' %}
     {{ knp_menu_render('sylius_admin.main', {'template': '@SyliusAdmin/shared/crud/common/sidebar/menu/menu.html.twig', 'currentClass': 'active'}) }}
```
And then this menu could be moved to a separate hookable, let's name it `'sylius_admin.common.component.sidebar.menu.items'`

Rafikooo replied:
This way we don’t break the RWD since both hookables are wrapped in the Bootstrap class.

crydotsnake replied:
Good idea! I implemented it, and its working great 🙂

crydotsnake replied:
But i'm unsure about the `priority`. Can we use `0` in this case or could that become problematic at some point?

Rafikooo replied:
Good catch. I found this entry: https://docs.sylius.com/the-customization-guide/customizing-templates#priority

I would be way too optimistic if I said it’s easy to find 🙂  
Also, we don’t have recommendations for Twig hooks written down in our docs yet, so it would be great to propose some.

Rafikooo replied:
And since we already mention the priority, please adjust it in the PR before we merge.

Rafikooo commented on src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/common/component/sidebar.yaml:
```diff
@@ -15,3 +15,11 @@ sylius_twig_hooks:
+                priority: 200
+            items:
+                template: '@SyliusAdmin/shared/crud/common/sidebar/menu/items.html.twig'
+                priority: 100
```
```suggestion
                priority: 0
```

Rafikooo commented on src/Sylius/Bundle/AdminBundle/Resources/config/app/twig_hooks/common/component/sidebar.yaml:
```diff
@@ -15,3 +15,11 @@ sylius_twig_hooks:
+        'sylius_admin.common.component.sidebar.menu':
+            search:
+                template: '@SyliusAdmin/shared/crud/common/sidebar/search.html.twig'
+                priority: 200
```
```suggestion
                priority: 100
```
